### ISSUE TITLE
Evaluate additional syntax coloring for Target Editor

### ISSUE THREAD

**Issue by laeubi** [eclipse-m2e/m2e-core#555]
Currently only the PDE location types have complete syntax coloring, the maven one is lacking coloring for its attributes and elements:

![grafik](https://user-images.githubusercontent.com/1331477/152482541-39f6916d-ff63-4e69-8865-8d7f22bb534d.png)

even some elements are marked brown here as the seem to match an attribute name of PDE IU location.

@mickaelistria can you give some hints how we can contribute to the PDE editor to support this?

**Comment by akurtakov** [comment]
Code is in https://git.eclipse.org/c/pde/eclipse.pde.ui.git/tree/ui/org.eclipse.pde.genericeditor.extension . It's very naive code https://git.eclipse.org/c/pde/eclipse.pde.ui.git/tree/ui/org.eclipse.pde.genericeditor.extension/src/org/eclipse/pde/internal/genericeditor/target/extension/reconciler/presentation/TargetPlatformPresentationReconciler.java#n91

**Comment by HannesWell** [comment]
Maybe the Syntax coloring should be switched off for location types unknown to PDE?


In general I'm wondering if https://github.com/eclipse/lemminx could also be applied to PDE's Target-Editor?
@mickaelistria would that be useful? We had a discussion a while ago that the Target-Editor should be improved and should have the same tooling support like the Target-Platform-DSL and I had the idea that the lemminx project could help here. What do you think? Would it be possible with lemminx to reference to other language definitions in parts of a text file, so that for Maven targets the Maven-language definition could be reused while for the IU-Locations PDE provides its own definition?
(Sorry for the off-topic).

**Comment by laeubi** [comment]
> Would it be possible with lemminx to reference to other language definitions in parts of a text file

I think `lemminx`is for generic xml, and `lemminx-maven` for the maven specific parts so we probably need a `lemminx-target` here :-)

**Comment by mickaelistria** [comment]
TL;DR: work needs to happen on PDE to be more tolerant. Closing.

The syntax highlighting in PDE Target is a custom one. See the link @akurtakov provided earlier. This reconciler could be improved to better handle unknown target types as well and default to something that would be like plain XML. I'm pretty sure there are already available plain XML highlighters available in Platform (for things like .project) or JDT (.classpath) or PDE (plugin.xml, feature.xml...) that we could use in place of a custom one here.

We could also imagine just relying on TM4E here, but it would then bring an additional dependency to PDE. Not sure it's worth it.

**Comment by laeubi** [comment]
> Closing

But why? We provide a new target type so it would make sense to also provide coloring for that type as well (even though PDE might need to first ad support for such thing).

**Comment by mickaelistria** [comment]
What kind of coloring are you (realistically) willing to achieve here that cannot be implemented in PDE directly by just doing plain XML stuff?

**Comment by laeubi** [comment]
For example we can embedd BND instructions that could need some custom coloring, I could think about coloring excluded items with some other color and so on.

So this is also kind of how/if could an generic editor be made extend beyond a single document (I have seen that actually there is some concept of "document parts" that could have different content types). If that's available for colors we probably could use the same for code-completions.

Or we could further elaborate if it would make sense to have a `lemminx-target` and PDE drops custom implementation and simply use a language server here as well...

**Comment by mickaelistria** [comment]
A language-server is only useful if it has multiple clients. I hardly see people needing full support for .target edition in other IDEs. the .target file has no XML schema, all its parsing/tooling are hardcoded bits in PDE and that allows to share many definitions between runtime and tool. Overall, for that case, a language-server wouldn't really make anything better here.
If we want plain XML highlighting in PDE target editor, we can just use one of the plain XML highlighers already available in PDE; or if there is none, just remove highlighter from the editor and let the default one from TM4E shining in.
I believe Generic Editor allow multiple participant in syntax highlighting, so one may already be able to add syntax highlighting or other form of assistance for BND instructions inside .target.

**Comment by laeubi** [comment]
But shouldn't lemminx already add xml support? So it seems good to reuse it here. I can eve write an XSD if necessary ...

**Comment by akurtakov** [comment]
Technically Platform and thus PDE as part of the same build can not depend on wwd/lemminx-* as that would create circular dependency. With that said - it's plain xml file so extensions to generic editor provided by other plugins should give smartness e.g. with WWD installed I get all autocompletion and even linked editing of opening/closing task. So if wwd/lemminx-* adds some smartness it will be there. Only issues is highlighting which I'm not deep into to have a say.

**Comment by akurtakov** [comment]
@mickaelistria Technically, this one is solved now but if you have suggestion for a "clean" way to move highlight/completion to m2e itself let me know.

**Comment by laeubi** [comment]
@akurtakov thanks for taking care of this!

**Comment by mickaelistria** [comment]
The .target editor is using the generic editor IIRC, so it should be possible to add the additional reconciler and completionProcessors as extensions to the generic editor attached to the .target content-type. The generic editor allows multiple "providers" for most features, so m2e can contribute the maven bits and the editor will take care of invoking it as well.
